### PR TITLE
fix issue when backward after forward(is_train=false)

### PR DESCRIPTION
**Pull request by lihaofd**
## Description ##
In this PR, it fixed issue about https://github.com/apache/incubator-mxnet/issues/13264
@pengzhao-intel, @TaoLv , @ciyongch 

### New features ###
- When requesting gradients during inference, since there is no allocated space and report warning rather than crash.

### Unit-test changes ###
- Add testcase robust_check_backward_after_inference in tests/python/unittest/test_operator.py for making sure there is no crash in this case.

## Checklist ##
 - [X] Passed code style checking (make lint).
 - [X] All changes have test coverage.
 - [ ] Code is well-documented.

### CODE REVIEW

**Comment by pengzhao-intel** [comment]
This is not a correct usage from the script. 
Is it possible to add a test case?

**Comment by taliesinb** [comment]
@lihaofd is this really the correct fix? It is completely valid behavior to request gradients when not in training mode, so I don't see why we should produce a warning at all. See https://mxnet.incubator.apache.org/api/python/executor/executor.html#mxnet.executor.Executor.backward which says "Note that in rare cases you want to call backward with is_train=False to get gradient during inference."

**Comment by pengzhao-intel** [comment]
@taliesinb just want to understand the real usage model, in what scenario we need this kind of feature?

And I think the problem in here is that we should set `is_train=True` in the forward path and then call backward.
https://mxnet.incubator.apache.org/api/python/executor/executor.html#mxnet.executor.Executor.forward

> is_train (bool, optional) – Whether this forward is for evaluation purpose. If True, a backward call is expected to follow.

Feel free to correct me :)

**Comment by taliesinb** [comment]
@pengzhao-intel Let's say you want to obtain gradients of an RNN model on an input sequence, for teaching, demonstration, or for model interpretability purposes (e.g. see which parts of the input sequence contributed to the classification). But your RNN model uses dropout. You obviously do not want to apply the dropout during inference, so you must set `is_train=False` for both the forward and backward pass. But with the current implementation of the RNN layer, you will get a crash/error if you do this. 

So here is my understanding of what the history of this is:

The previous design of Forward and Backward confused these two questions of "do i want gradients" and "am i training" together into one thing, even though they are actually separate. Recently piiswrong improved the situation by adding the "am i training" option to the the Backward pass, so that you could give `is_train=False` to indicate that you are getting gradients but in inference mode. He added a note to the documentation for Backward that explicitly talks about this. I quoted that note above, but here it is again:

> Note that in rare cases you want to call backward with is_train=False to get gradient during inference."

But you *also* have to apply `is_train=False` during Forward pass otherwise things like dropout will be applied. So piiswrong seems to have forgotten to update the note in the documentation for Forward, the one *you* quoted, which still implies the now-incorrect idea that "i want gradients" = "i am training". 

Other than one documentation problem, the design he intended seems straightforward: if you want to get gradients in inference mode, you must pass `is_train=False` to both forward and backward.

The bug I reported is that RNN is not conforming to that design. We can also fix the documentation for Forward to say:

> is_train (bool, optional) – True if this forward pass is for training purposes, false if it is for inference purposes. If True, training-time techniques like dropout will be applied and batch normalization running estimates will be updated. A call to Backward can always be performed after the call to Forward, but the value for the is_train option should match (both True or both False).  

Am I misunderstanding something? I do not see another way to understand what piiswrong was trying to do.

**Comment by pengzhao-intel** [comment]
@taliesinb really thanks for the background introduction. I actually don't know these before and it's really helpful for us. 

We need to re-evaluate how it affects the current design and how to improve it. 

@Taolv @ciyongch @piiswrong @szha

**Comment by TaoLv** [comment]
So not only RNN, many other operators will also be affected by this assumption change.

**Comment by taliesinb** [comment]
@pengzhao-intel no problem! 

Here is some more evidence that the is_train flag is not intended to be used by layers to decide whether they need to compute gradients:

```
void GraphExecutor::RunOps(bool is_train, size_t topo_start, size_t topo_end) {
  // Update context
  const auto& idx = graph_.indexed_graph();
  for (size_t nid = topo_start; nid < topo_end; ++nid) {
    OpNode& opnode = op_nodes_[nid];
    if (opnode.skip_exec_node) continue;
    const auto& inode = idx[nid];
    if (inode.source->is_variable()) continue;
    opnode.exec->op_ctx.is_train = is_train;
    opnode.exec->op_ctx.need_grad = need_grad_;
  }
```
Notice that the `need_grad_` flag is separate from the `is_train` flag. It turns out that `need_grad_` is set in `InitFullGraph` based on whether any gradient arrays had a non-null request type. So that's what we need!

Therefore the solution I think for the RNN bug straightforward: the check `if (ctx.is_train)` on line 417 of `rnn-inl.h` hould be replaced by the check `if (ctx.need_grad)`

Similar replacements can be made elsewhere too.

**Comment by taliesinb** [comment]
If there is no objection, can I create a PR that makes this switch consistently across the codebase? I should be able to have something in a few hours.

**Comment by pengzhao-intel** [comment]
@taliesinb thanks, please go ahead. I will ping you when we get a solution to handle this situation.

**Comment by taliesinb** [comment]
@pengzhao-intel 

Here's my branch: https://github.com/taliesinb/mxnet/tree/bugfix/use_need_grad

I need to make a JIRA ticket first, and I need to add things like tests, of course. I also haven't hooked this up in the Python API, yet, either.

**Comment by taliesinb** [comment]
I'm unlikely to have time to keep working on this... but I can give the short version of what I figured out by working on my branch.  

Basically, we will need an additional ForwardEx API function in the C api that takes a second boolean which is whether gradients are requested or not. So you call ForwardEx four different ways: 

1. `ForwardEx(is_train=false, needs_grad=false)`: ordinary evaluation
2. `ForwardEx(is_train=false, needs_grad=true)`: will call Backward later, but don't turn on dropout behavior or modify batchnorm moments etc)
3. `ForwardEx(is_train=true, needs_grad=true)`: ordinary training
4. `ForwardEx(is_train=true, needs_grad=false)`: ordinary evaluation, but wish to see dropout behavior etc (for example, doing monte carlo stuff with dropout layer)

Then, all through the system these two flags would need to be separately tracked and checked appropriately. Mostly this is fairly simple: layers should check the needs_grad flag in RunContext to decide whether to allocate storage / save state for gradients, and check the is_train flag for whether to use training behavior for dropout, leakyrelu, etc.

But for batch norm it is a little complex, as those behaviors are currently totally mixed up. 

Also, the subgraph executors use the need_grad flag from the RunContext and use it to set the is_train flag of the subgraph's Forward pass, which is quite confusing and definitely wrong. After the refactor I'm proposing, the subgraph executor will need to pass each flag on separately to the subgraph's Forward pass.

I couldn't finish this because its a substantial refactor that needs very careful testing, and also modifying the python interface, gluon, etc. Probably best done by someone who works fulltime on mxnet, and after some discussion as well.

**Comment by pengzhao-intel** [comment]
@taliesinb understand, there're lots of works to make this change happen. 
Thanks all of your great inputs again.
